### PR TITLE
finished CRUD operations with users

### PR DESCRIPTION
**Pull request by IdanPhilosoph**
finished login, register and edit profile for users in the server

### CODE REVIEW

File: backend/.env
Comment (Elon-Abulafia) [inline backend/.env:1]: As a rule of thumb, never upload your `.env` file into your git repo.

File: backend/server.py [hunk 1]
```diff
@@ -1,18 +1,31 @@
...
+    'password': os.environ.get("DATABASE_PASSWORD"),
+    'database': os.environ.get("DATABASE_NAME"),
+}
+
 class User(UserMixin):
     def __init__(self, id, name, email, password):
         self.id = id
         self.name = name
         self.email = email
         self.password = password
```
Comment (Elon-Abulafia) [inline backend/server.py]: This doesn't need to be in this file.

File: backend/server.py [hunk 2]
```diff
@@ -28,17 +41,11 @@ def load_user(email):
     # Close the cursor and connection
     cursor.close()
     connection.close()
```
Comment (Elon-Abulafia) [inline backend/server.py]: Consider moving the entire DB interaction logic to a class that handles the connection, cursor, etc... (in a different file)

File: backend/server.py [hunk 3]
```diff
@@ -28,17 +41,11 @@ def load_user(email):
...
+
     if user:
         return User(user[2], *user[1:])
     return None
```
Comment (Elon-Abulafia) [inline backend/server.py]: What exactly is `user[2]` or `*user[1:]`?
You shouldn't write code with magic numbers that don't mean anything.
Also every function should have just a single `return` statement.

File: backend/server.py [hunk 4]
```diff
@@ -118,12 +125,13 @@ def login():
...
         return error_response('Invalid email or password', 401)
 
+# Route for handling edit_profile requests
 @app.route('/edit_profile', methods=['POST'])
 @login_required
 def edit_profile():
     user_data_json = request.get_json()
 
-    # Check if the user with the given email already exists
+    # Check if the new email is already in use by another user
```
Comment (Elon-Abulafia) [inline backend/server.py]: Leaving comments in your code like that isn't recommended.
If you want to document your functions, leave docstrings at the definition of the function.

File: backend/server.py [hunk 5]
```diff
@@ -144,4 +152,6 @@ def edit_profile():
     return success_response(message='Profile updated successfully')
 
 if __name__ == '__main__':
+    
+    # run the server
```
Comment (Elon-Abulafia) [inline backend/server.py]: Again, pointless comment.

File: backend/server.py [hunk 6]
```diff
@@ -118,12 +125,13 @@ def login():
...
 @app.route('/edit_profile', methods=['POST'])
 @login_required
 def edit_profile():
     user_data_json = request.get_json()
 
-    # Check if the user with the given email already exists
+    # Check if the new email is already in use by another user
     if current_user.email != user_data_json['email'] and user_exists(user_data_json['email']):
         print(f"User with email {user_data_json['email']} already exists.")
         return error_response('email is already in use', 401)
```
Comment (Elon-Abulafia) [inline backend/server.py]: Maybe try to handle logged in users in a different form.
This doesn't really work with multiple users.

**Review by Elon-Abulafia — CHANGES_REQUESTED** [review]

File: backend/server.py (hunk 6 above)
Comment (IdanPhilosoph) [reply backend/server.py]: this shouldnt be an issue and therefore havent been fixed

**Comment by IdanPhilosoph** [comment]
fixed all of the issues from the CR

File: backend/database.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
+from dotenv import load_dotenv
+import mysql.connector
+from user import User
+import os
+
+# Load environment variables from .env file
+load_dotenv()
+
+# DB config used for accessing the database
+db_config = {
```
Comment (IdanPhilosoph) [inline backend/database.py]: You should have consts file where you will keep the DB host & Database

File: backend/database.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    'host': os.environ.get("HOST"),
+    'user': os.environ.get("DATABASE_USERNAME"),
+    'password': os.environ.get("DATABASE_PASSWORD"),
+    'database': os.environ.get("DATABASE_NAME"),
+}
+
+def get_user_by_email(email):
```
Comment (IdanPhilosoph) [inline backend/database.py]: Create a DB connection class, singleton, and all the access methods will go through that class

File: backend/database.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    connection = mysql.connector.connect(**db_config)
+    cursor = connection.cursor()
+
+    select_query = "SELECT * FROM users WHERE email = %s"
+    cursor.execute(select_query, (email,))
+    user = cursor.fetchone()
```
Comment (IdanPhilosoph) [inline backend/database.py:12]: Check what happens when `fetchone()` returns nothing

File: backend/database.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    if user:
+        column_names = [column[0] for column in cursor.description]
+        user_dict = {column: value for column, value in zip(column_names, user)}
+
+        cursor.close()
+        connection.close()
+
+        # email is used for the id as it is the primary key
+        return User(user_dict["email"], user_dict["name"], user_dict["email"], user_dict["password"])
+    else:
```
Comment (IdanPhilosoph) [inline backend/database.py]: `else` is redundent

File: backend/database.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        return None
+
+def user_exists(email):
```
Comment (IdanPhilosoph) [inline backend/database.py]: redundent

File: backend/database.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    connection = mysql.connector.connect(**db_config)
+    cursor = connection.cursor()
+
+    check_query = "SELECT * FROM users WHERE email = %s"
+    cursor.execute(check_query, (email,))
+    existing_user = cursor.fetchone()
```
Comment (IdanPhilosoph) [inline backend/database.py]: exists instead of fetchone(?)

File: backend/database.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+    return existing_user is not None
+
+def add_user(name, email, password):
+
+    connection = mysql.connector.connect(**db_config)
+    cursor = connection.cursor()
+
+    insert_query = "INSERT INTO users (name, email, password) VALUES (%s, %s, %s)"
+    user_data = (name, email, password)
```
Comment (IdanPhilosoph) [inline backend/database.py:49]: Not necessary

File: backend/server.py [hunk 14]
```diff
@@ -70,76 +32,39 @@ def error_response(message, status_code):
 @app.route('/register', methods=['POST'])
 def register():
     user_data_json = request.get_json()
```
Comment (IdanPhilosoph) [inline backend/server.py]: `get_json()` is deprecated, use `json()` instead.

File: backend/server.py [hunk 15]
```diff
@@ -70,76 +32,39 @@ def error_response(message, status_code):
...
-    connection.close()
-
     return success_response(message='Registration successful')
 
 # Route for handling login requests
 @app.route('/login', methods=['POST'])
 def login():
-    
     login_data_json = request.get_json()
+    user = get_user_by_email(login_data_json['email'])
```
Comment (IdanPhilosoph) [inline backend/server.py]: change PK to id, email is expensive to work with as a PK

**Pull request merged**